.pre_commit/generate-facets/generate.py

```diff
@@ -0,0 +1,256 @@
+# Copyright 2018-2024 contributors to the OpenLineage project
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import copy
+import json
+import logging
+import os
+import pathlib
+import re
+import shutil
+import tempfile
+from collections import defaultdict
+
+import click
+from datamodel_code_generator import DataModelType, PythonVersion
+from datamodel_code_generator.imports import Import
+from datamodel_code_generator.model import get_data_model_types
+from datamodel_code_generator.parser.jsonschema import JsonSchemaParser
+from datamodel_code_generator.types import Types
+
+log = logging.getLogger(__name__)
+
+def camel_to_snake(string: str) -> str:
+    """Convert camel_case string to SnakeCase"""
+    _under_score_1 = re.compile(r"([^_])([A-Z][a-z]+)")
+    _under_score_2 = re.compile("([a-z0-9])([A-Z])")
+    subbed = _under_score_1.sub(r"\1_\2", string)
+    return _under_score_2.sub(r"\1_\2", subbed).lower()
+
+
+# import attrs instead of dataclass
+DATA_MODELS_TYPES = get_data_model_types(
+    DataModelType.DataclassesDataclass, target_python_version=PythonVersion.PY_38
+)
+NEW_MODEL = DATA_MODELS_TYPES.data_model
+NEW_MODEL.DEFAULT_IMPORTS = (
+    Import.from_full_path("attr.define"),
+    Import.from_full_path("attr.field"),
+)
+
+# locations definitions
+PARENT_LOCATION = pathlib.Path(__file__).resolve().parent.parent.parent
+BASE_SPEC_LOCATION = PARENT_LOCATION / "spec" / "OpenLineage.json"
+FACETS_SPEC_LOCATION = PARENT_LOCATION / "spec" / "facets"
+DEFAULT_OUTPUT_LOCATION = PARENT_LOCATION / "client" / "python" / "openlineage" / "client" / "generated"
+TEMPLATES_LOCATION = pathlib.Path(__file__).resolve().parent / "templates"
+PYTHON_CLIENT_LOCATION = PARENT_LOCATION / "client" / "python"
+
+# custom code
+SET_PRODUCER_CODE = """
+PRODUCER = DEFAULT_PRODUCER
+
+def set_producer(producer: str) -> None:
+    global PRODUCER  # noqa: PLW0603
+    PRODUCER = producer
+"""
+HEADER = (
+    "# Copyright 2018-2024 contributors to the OpenLineage project\n"
+    "# SPDX-License-Identifier: Apache-2.0\n\n"
+)
+
+# structures to customize code generation
+REDACT_FIELDS = {
+    "Assertion": ["column"],
+    "InputField": ["namespace", "name", "field"],
+    "Dataset": ["namespace", "name"],
+    "InputDataset": ["namespace", "name"],
+    "OutputDataset": ["namespace", "name"],
+    "Job": ["namespace", "name"],
+    "Run": ["runId"],
+    "RunEvent": ["eventType", "eventTime"],
+    "NominalTimeRunFacet": ["nominalStartTime", "nominalEndTime"],
+    "ParentRunFacet": ["job", "run"],
+    "SourceCodeLocationJobFacet": ["type", "url"],
+    "DatasourceDatasetFacet": ["name", "uri"],
+    "OutputStatisticsOutputDatasetFacet": ["rowCount", "size"],
+    "SourceCodeJobFacet": ["language"],
+    "ErrorMessageRunFacet": ["programmingLanguage"],
+}
```

Comment:
Well, I could put it into some structured file, e.g. yaml, document the usage and load it from there. Would that work better?

Reply:
I'm not sure where the file should land. In `client/python`?

Reply:
Not sure how yaml would look like or what would you want to achieve. I think better idea would be validation via pre-commit - for example, if we need to add something for all new facets here, the pre-commit could check if the facet is mentioned here.

Reply:
Not every facet has to have unredacted fields. Unless we force putting here facets with empty list just so they are, it wouldn't work.

Reply:
>Unless we force putting here facets with empty list just so they are, it wouldn't work.

TBH, that's not a bad idea - it would force people to think about redacting fields when adding the facets.

Reply:
:+1:

Reply:
I changed the structure of the file - now it allows you to define redact fields on module & class level (classes repeat their names).
I also added pre-commit check that alerts you if you add new class/facet generated but leave it without any information in the `redact_fields.yml`.